swarm/benches/connection_handler.rs

```diff
@@ -0,0 +1,368 @@
+        SpinningBehaviour10::bench().name(j).poll_count(10000).protocols_per_behaviour(100),
+    ];
+    lucky_20::[
+        SpinningBehaviour20::bench().name(k).poll_count(5000).protocols_per_behaviour(1),
+        SpinningBehaviour20::bench().name(l).poll_count(5000).protocols_per_behaviour(10),
+        SpinningBehaviour20::bench().name(m).poll_count(5000).protocols_per_behaviour(100),
+    ];
+}
+//fn main() {}
+
+trait BigBehaviour: Sized {
+    fn behaviours(&mut self) -> &mut [SpinningBehaviour];
+
+    fn for_each_beh(&mut self, f: impl FnMut(&mut SpinningBehaviour)) {
+        self.behaviours().iter_mut().for_each(f);
+    }
+
+    fn any_beh(&mut self, f: impl FnMut(&mut SpinningBehaviour) -> bool) -> bool {
+        self.behaviours().iter_mut().any(f)
+    }
+
+    fn run_bench(
+        c: &mut Criterion,
+        protocols_per_behaviour: usize,
+        spam_count: usize,
+        static_protocols: bool,
+    ) where
+        Self: Default + NetworkBehaviour,
+    {
+        let mut swarm_a = new_swarm(Self::default());
+        let mut swarm_b = new_swarm(Self::default());
+
+        let behaviour_count = swarm_a.behaviours().len();
+        let protocol_count = behaviour_count * protocols_per_behaviour;
+        let protocols = (0..protocol_count)
+            .map(|i| {
+                if static_protocols {
+                    StreamProtocol::new(format!("/protocol/{i}").leak())
+                } else {
+                    StreamProtocol::try_from_owned(format!("/protocol/{i}")).unwrap()
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let mut protocol_chunks = protocols.chunks(protocols_per_behaviour);
+        swarm_a.for_each_beh(|b| b.protocols = protocol_chunks.next().unwrap().to_vec());
+        let mut protocol_chunks = protocols.chunks(protocols_per_behaviour);
+        swarm_b.for_each_beh(|b| b.protocols = protocol_chunks.next().unwrap().to_vec());
+
+        swarm_a.for_each_beh(|b| b.iter_count = spam_count);
+        swarm_b.for_each_beh(|b| b.iter_count = 0);
+
+        swarm_a.for_each_beh(|b| b.other_peer = Some(*swarm_b.local_peer_id()));
+        swarm_b.for_each_beh(|b| b.other_peer = Some(*swarm_a.local_peer_id()));
+
+        static mut OFFSET: usize = 0;
+        let offset = unsafe {
+            OFFSET += 1;
+            OFFSET
+        };
+
+        swarm_a
+            .listen_on(format!("/memory/{offset}").parse().unwrap())
+            .unwrap();
+        swarm_b
+            .dial(format!("/memory/{offset}").parse::<Multiaddr>().unwrap())
+            .unwrap();
+
+        let name = format!(
+            "{}::bench().poll_count({}).protocols_per_behaviour({})",
+            std::any::type_name::<Self>(),
+            spam_count,
+            protocols_per_behaviour
+        );
+
+        c.bench_function(&name, |b| {
+            swarm_a.for_each_beh(|b| b.finished = false);
+            swarm_b.for_each_beh(|b| b.finished = false);
+            b.iter(|| {
```

Comment:
Have you considered using some of the other timing loops to deal with the setup? https://bheisler.github.io/criterion.rs/book/user_guide/timing_loops.html

I think this might allow us to remove the `restarting` state within the `SpinningBehaviour`?

Reply:
done